src/Nethermind/Nethermind.State/PreBlockCaches.cs

```diff
@@ -4,34 +4,124 @@
+        _clearStorageCache = () => ClearConcurrentCache(ref _storageCache);
+        _clearStateCache = () => ClearConcurrentCache(ref _stateCache);
+        _clearRlpCache = () => ClearConcurrentCache(ref _rlpCache);
+        _clearPrecompileCache = () => ClearConcurrentCache(ref _precompileCache);
+    }
+
+    public ConcurrentDictionary<StorageCell, byte[]> StorageCache => _storageCache;
+    public ConcurrentDictionary<AddressAsKey, Account> StateCache => _stateCache;
+    public ConcurrentDictionary<NodeKey, byte[]?> RlpCache { get; } = new(LockPartitions, InitialCapacity);
+    public ConcurrentDictionary<PrecompileCacheKey, (ReadOnlyMemory<byte>, bool)> PrecompileCache { get; } = new(LockPartitions, InitialCapacity);
+
+    public async Task ClearCachesInBackground()
+    {
+        Task t0 = Task.CompletedTask;
+        Task t1 = Task.CompletedTask;
+        Task t2 = Task.CompletedTask;
+        Task t3 = Task.CompletedTask;
+
+        bool isDirty = false;
+        if (!_storageCache.IsEmpty)
+        {
+            isDirty = true;
+            t0 = Task.Run(_clearStorageCache);
+        }
+        if (!_stateCache.IsEmpty)
+        {
+            isDirty = true;
+            t1 = Task.Run(_clearStateCache);
+        }
+        if (!_rlpCache.IsEmpty)
+        {
+            isDirty = true;
+            t2 = Task.Run(_clearRlpCache);
+        }
+        if (!_precompileCache.IsEmpty)
+        {
+            isDirty = true;
+            t3 = Task.Run(_clearPrecompileCache);
+        }
+
         if (isDirty)
         {
-            StorageCache.Clear();
-            StateCache.Clear();
-            RlpCache.Clear();
+            await Task.WhenAll(t0, t1, t2, t3);
+        }
+    }
+
+    public bool ClearImmediate()
+    {
+        bool isDirty = false;
+        if (!_storageCache.IsEmpty)
+        {
+            isDirty = true;
+            ClearConcurrentCache(ref _storageCache);
+        }
+        if (!_stateCache.IsEmpty)
+        {
+            isDirty = true;
+            ClearConcurrentCache(ref _stateCache);
+        }
+        if (!_rlpCache.IsEmpty)
+        {
+            isDirty = true;
+            ClearConcurrentCache(ref _rlpCache);
+        }
+        if (!_precompileCache.IsEmpty)
+        {
+            isDirty = true;
+            ClearConcurrentCache(ref _precompileCache);
         }
 
         return isDirty;
     }
 
+    private void ClearConcurrentCache<TKey, TValue>(ref ConcurrentDictionary<TKey, TValue> cache, [CallerArgumentExpression(nameof(cache))] string? paramName = null)
+    {
+        cache.NoResizeClear();
+
+        if (!cache.IsEmpty)
+        {
+            // Guard in case the cache is not empty after the loop
+            if (_logger.IsWarn)
+            {
+                _logger.Warn($"{paramName} didn't empty. Purging.");
+            }
+            // Recreate rather than clear to avoid resizing
+            cache = new(LockPartitions, InitialCapacity);
+        }
```

Comment:
If you do locking in `NoResizeClear` this is not needed.

Reply:
Done